### DISCUSSION TITLE
How to set up remote access with cockpit and nginx

### DISCUSSION THREAD

**Discussion by jamminjames** [cockpit-project/cockpit#19653] [Q&A] [answered]
I am trying to get the cockpit web console working with a remote connection on our server following [these instructions](https://cockpit-project.org/external/wiki/Proxying-Cockpit-over-NGINX). Right now, I can log in remotely only with an HTTP connection using the ip and 9090 port. I want to connect via HTTPS, and using the domain name instead of the ip would be nice, but I'm not sure if it's necessary or even preferable.

I'm confused trying to follow the instructions. Where it says to create a virtual server block, it gives an example line of `server_name    cockpit.domain.tld;` Is "domain.tld" a stand-in for your actual domain, or should I use it exactly like that? 

The instructions also say to put this in the the Cockpit config file:
```
[WebService]
Origins = https://cockpit.domain.tld wss://cockpit.domain.tld
ProtocolHeader = X-Forwarded-Proto
```
... so would the nginx server block be referring to that, or should I use our actual domain there as well?

If it is our domain, and we're setting up a subdomain name of 'cockpit', does that need to be set up in our DNS records as well?

And as for the certificate, can I just copy over our website's SSL cert, or do I need to come up with a new one?

Thanks for any help!

**Comment by martinpitt** [comment]
`cockpit.domain.tld` is a stand-in, i.e. please replace it with your actual server name and domain (FQDN). Perhaps it should say `cockpit.example.com`?

Wrt. the certificate: The whole idea is that nginx does the reverse proxying and TLS. So from the outside, the browser talks to nginx on the configured domain name. nginx then forwards the requests to cockpit on (usually) http://localhost:9090, so cockpit itself does *not* speak TLS or even need a certificate. You usually use the same TLS cert for your entire site, including any subdomains (like cockpit.*).

However, cockpit needs to be told what domain name it "appears" to have, so that it can compute correct links and enforce a correct "same origin" policy. It cannot detect that by itself. That's what you do with `Origins =`.

Does that make sense?

**Reply by jamminjames** [reply]
Thanks for your help. Okay, so I've set it up like the example I linked to, and replaced the stand-in with our domain name, and created the cockpit.conf file with the Origins line as they show, `Origins = https://cockpit.domain.tld wss://cockpit.domain.tld` (with our domain). I'm pointing to our website's TLS cert in the server block the instructions say to set up. 

It says I should be able to access Cockpit now via `https://cockpit.domain.tld` (with our domain name), but it's not working, it just goes to another subdomain we have. I thought it might have to do with our CDN, Cloudflare. So I did set up a CNAME there for cockpit.domain.tld, but no change after several hours. I don't think the issue there is that it needs more time to propagate, because at first I got a 'no such address' type error.

I also cannot use https://domain.tld:9090 (timeout error "This site can’t be reached") nor the ip with the port (I get NET::ERR_CERT_AUTHORITY_INVALID). With the latter, I could just go ahead through the warning, but I'd like it to work without that and be secure.

**Reply by martinpitt** [reply]
https://domain.tld:9090/ should not ever work, unless your webserver (or something else) is also listening there. You should either firewall off this port to the outside, or configure cockpit-ws to listen to 127.0.0.1 only: https://cockpit-project.org/guide/latest/listen.html#listen-systemd

"just goes to another subdomain" sounds like your nginx config is missing something to actually proxy the connection to http://localhost:9090 (i.e. to cockpit)? Check the server logs and possible typos in the config?

**Reply by jamminjames** [reply]
This is the nginx server block I have for cockpit, following the instructions I linked to earlier. What do you think?
```
server {
    listen         80;
    listen         443 ssl;
    server_name    cockpit.humortimes.com;

    # SSL
    ssl_certificate_key /path/to/keyfile.key;
    ssl_certificate /path/to/certfile.pem;
    ssl_trusted_certificate /path/to/geochain.pem;

    location / {
        # Required to proxy the connection to Cockpit
        proxy_pass https://127.0.0.1:9090;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-Proto $scheme;

        # Required for web sockets to function
        proxy_http_version 1.1;
        proxy_buffering off;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";

        # Pass ETag header from Cockpit to clients.
        # See: https://github.com/cockpit-project/cockpit/issues/5239
        gzip off;
    }
}
```

**Reply by martinpitt** [reply]
This looks fine at first sight. Sorry, I don't have any other idea what could be wrong now.

**Reply by jamminjames** [reply]
Elsewhere I'm using fastcgi_pass (to the php-fpm.sock), will using proxy_pass here cause any conflict?

Also, I have `gzip on` for the site, will having `gzip off` here only affect this server block, or could that conflict with the other directive?

**Reply by jamminjames** [reply]
A redirect trace shows it's getting redirected to our main domain, showing "X-Redirect-By: WordPress".

However, for it to get to the main domain, since the DNS CNAME file is for cockpit.humortimes.com, it must go through the above .conf file first, right? And the conf file is supposed to proxy it to the localhost 9090 port. So why does it proceed to the main domain? Does this file need tweaking when a CDN like Cloudflare is involved? Maybe the `proxy_set_header` lines need to be changed? I'm not sure what they do.

How can I troubleshoot this issue?

**Reply by martinpitt** [reply]
I'm afraid I have no idea how wordpress enters the game here. Maybe that configuration "grabs" all other virtual domains somehow?

**Reply by jamminjames** [reply]
But what about the CDN (Cloudflare) we're using. Is there something we need to do with our Cockpit setup differently in the case of a CDN?

**Reply by martinpitt** [reply]
I've never set up cloudflare, I have no idea. My hunch is that it is unrelated, and it's your wordpress configuration which "grabs" the cockpit domain.

**Reply by jamminjames** [reply]
CDNs are very common though. Is there anyone with this cockpit project that might know for sure if I need to tweak the cockpit server block to deal with the CDN? Again, I don't see how Wordpress could 'grab' it first, it is in the main domain, and the DNS points to cockpit.mywebsite.com.

**Comment by jamminjames** [comment, marked as answer]
OKAY! Finally figured it out. It was simply that my main server block was more specific than the subdomain server block. It had the IP and the port, whereas the subdomain block only had the port! Nginx will pick the more specific one first, as the request is coming to that IP.

Now it’s working. Thanks for your help and your patience!

**Reply by martinpitt** [reply]
I'm glad you figured it out!
